src/compiler/checker.ts

```diff
@@ -4583,7 +4583,10 @@ namespace ts {
                 let requireOptionalProperties = relation === subtypeRelation && !(source.flags & TypeFlags.ObjectLiteral);
                 for (let targetProp of properties) {
                     let sourceProp = getPropertyOfType(source, targetProp.name);
+
                     if (sourceProp !== targetProp) {
+                        // sourceProp !== targetProp -- ie: source and target have distinct declarations with the same name

```

Comment:
Okay, I see that in the example you gave above, the comparison occurs in checking `typesRelatedTo()` for the type arguments in the case where the LHS and RHS types of the assignment are both references pointing to the same target.

I removed the comment :).